src/ml/neural_net/combine_iterator.hpp

```diff
@@ -0,0 +1,147 @@
+/* Copyright © 2020 Apple Inc. All rights reserved.
+ *
+ * Use of this source code is governed by a BSD-3-clause license that can
+ * be found in the LICENSE.txt file or at
+ * https://opensource.org/licenses/BSD-3-Clause
+ */
+
+#ifndef ML_NEURAL_NET_COMBINE_ITERATOR_HPP_
+#define ML_NEURAL_NET_COMBINE_ITERATOR_HPP_
+
+#include <exception>
+#include <memory>
+#include <type_traits>
+
+#include <ml/neural_net/combine_base.hpp>
+
+namespace turi {
+namespace neural_net {
+
+template <typename T>
+class IteratorPublisher;
+
+/**
+ * Interface for objects that produce a sequence of values, using the
+ * conventional iterator interface.
+ *
+ * This type facilitates wrapping traditional iterator-style code as a
+ * Publisher.
+ */
+template <typename T>
+class Iterator : public std::enable_shared_from_this<Iterator<T>> {
+ public:
+  using Output = T;
+
+  virtual ~Iterator() = default;
+
+  /**
+   * Returns true as long as the underlying sequence contains more values.
+   *
+   * \todo If we have Optional<T>, we can remove this method and have Next()
+   *       instead return Optional<Output>.
+   */
+  virtual bool HasNext() const = 0;
+
+  /** Returns the next value in the sequence. May throw on error. */
+  virtual Output Next() = 0;
+
+  /** Returns a Publisher wrapping this Iterator. */
+  std::shared_ptr<IteratorPublisher<T>> AsPublisher() {
+    return std::make_shared<IteratorPublisher<T>>(this->shared_from_this());
+  }
+};
+
+/**
+ * Concrete Publisher that wraps an Iterator.
+ *
+ * The resulting Publisher is unicast: each iterated value will go only to
+ * whichever Subscriber triggered the iteration.
+ */
+template <typename T>
+class IteratorPublisher : public Publisher<T> {
+ public:
+  using Output = T;
+
+  explicit IteratorPublisher(std::shared_ptr<Iterator<Output>> iterator)
+      : iterator_(std::move(iterator)) {}
+
+  void Receive(std::shared_ptr<Subscriber<Output>> subscriber) override {
+    auto subscription =
+        std::make_shared<IteratorSubscription>(subscriber, iterator_);
+    subscriber->Receive(std::move(subscription));
+  }
+
+ private:
+  // All of the logic lives in the implementation of Subscription, which relies
+  // on the assumption that only one Subscription at a time will access the
+  // shared Iterator.
+  class IteratorSubscription : public Subscription {
+   public:
+    IteratorSubscription(std::shared_ptr<Subscriber<Output>> subscriber,
+                         std::shared_ptr<Iterator<Output>> iterator)
+        : subscriber_(std::move(subscriber)), iterator_(std::move(iterator)) {}
+
+    bool IsActive() const { return subscriber_ != nullptr; }
+
+    void Cancel() override { subscriber_.reset(); }
+
+    void Request(Demand demand) override {
```

Comment:
https://github.com/apple/turicreate/pull/3000/files#r382174687

I think the problem also comes from here. If it's canceled, the subscriber kept queueing its `promise` queue, while subscription doesn't reply but ignores its request.

If one canceled the subscription and used its subscriber to get a future inadvertently, the future of that will hang forever and it's hard to debug. I think it's better to explicitly reply something like `SubscriptionCancelled`.